### ISSUE TITLE
Bundle final bineary jar with low res maps and low res height data set

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#901]
**Is your feature request related to a problem? Please describe.**
- Currently, there's a choice of selecting high res vs. low res map by using MapMetaData's either one of the methods : 
```java
    String getHiResFile() {
        return hiResFilename;
    }

    public String getLoResFile() {
        return loResFilename;
    }
```
- However, in `MEGDRMapReader`, there's no alternative in loading the low res height data as it has been deleted from repo. 
- If players do not have the bandwidth or are offline, he will not be able to agree with downloading high res maps or the high res elevation data set during runtime and simulation will halt.

**Describe the rationale**
- Need a fallback mechanism to use low res surface map and height data set

**Describe the solution you desire to see**
- Designate those low res maps at mapdata submodule's `/resources/maps/` as the fallback, in the event that URL connection is not available for downloading high res maps, assuming that while there is no maps in home directory.
- Add `megt90n000cb.img` back to `/maps/`. It provides 
- If user cannot download `megt90n000eb.img`, fallback to `megt90n000cb.img`

**Related Classes**
- In `IntegerMapData`, 
```java
 	/**
 	 * Constructor.
 	 * 
	 * @param name   the name/description of the data
 	 * @param filename   the map data file name.
 	 * @throws IOException Problem loading map data
 	 */
 	IntegerMapData(MapMetaData newMeta) throws IOException {

		if (meta == null || !newMeta.getName().equals(meta.getName())) {
			this.meta = newMeta;
			// Load data files
			pixels = loadMapData(newMeta.getHiResFile());
			rho =  pixelHeight / Math.PI;
			logger.info("Loaded " + meta.getHiResFile() + " with pixels " + pixelWidth + "x" + pixelHeight + ".");
		}
 	}
```
- In `MEGDRMapReader`, 
```java
	/**
	 * Loads the elevation data into short array.
	 * 
	 * @return
	 */
	public short[] loadElevation() {	 
		
	    try (InputStream inputStream = new FileInputStream(FileLocator.locateFile(FILE))) {

			// Use ByteStreams to convert to byte array
			byte[] bytes = ByteStreams.toByteArray(inputStream);
			
			elevation = convertByteArrayToShortIntArray(bytes);

			
			height = (short) Math.sqrt(elevation.length / 2);
			width = (short) (height * 2);
			
	            
		} catch (Exception e) {
			 System.out.println("Problems in inputStream: " + e.getMessage());
		}
	    
        return elevation;
	}
```

**Specifications  (please complete the following information):**
 - pre 3.6.0 build 8128

**Additional context**
- `megt90n000cb.img` provides 
--	map resolution of 1440x720
--	4 pixels per degree (or 0.25 by 0.25 degrees)
--	map scale of 14.818 km per pixel
	
- `megt90n000eb.img` provides
--	map resolution of 5760x2880
--	16 pixels per degree (or 0.0625by 0.0625 degrees)
--	map scale of 3.705 km per pixel

**Comment by mokun** [comment]
@bevans2000 

How do you like to code the fallback mechanism ? 

Q1 : Will player have to answer many "no"'s to downloading high res map or height data ? Should this dialog box be popped up only once per simulation session fair enough ?

Q2 : Should we indicate somewhere that it's using low res map/height data and provide a button to click to download high res map ?

**Comment by bevans2000** [comment]
There are 2 issues here.

Firstly the LoRes/HiRes images in the ```MapData``` class are only used by the Navigator window. There we already prompt the user if they want to download a hires map. We just need to make sure that only the hires image of the default map layer is bundled in the JAR file to keep the size down. Nothing more. The LoRes images we can bundled as they are smaller if you want. I feel this is in a good place now and nothing more is needed. Maybe review which map images are bundled.
Nevertheless; the loading from the internet could be improved to do it asynchronously which will give a better user experience. I'll think about that.

Secondly the MEGDR reader uses a different way to handle the image files. It loads it all into memory and then steps into it for terrain calculations. I've already mentioned that we should not do this and read the values direct from file and cache the most recent/well used values. This is something we could do as this would remove the need to have the image loaded. This leads me to my suggestion.
The Terrain logic would run on the bundled low res map data initially. When the hires terrain data is downloaded; it could then flip to use that better terrain data. To achieve this we need to look-up the terrain details directly from file such that a new file can be "dropped" in underneath whilst the sim is running. However, we may choose to not persistently hold the terrain elevation data for a Settlement because, in this approach, it will be updated async in the background.
It this world we could bundle the low res MEGDR data but kick off a download of the hires data in the background. 
Remember, any downloaded files are reusable for the next run. They are only download once per installation.
To do this we would have to:

1. Make MEGDR reader (or the calling class) lookup elevation directly form the image file. And cache recent values.
2. Make the downloading of files async with some callback notification.
3. Change the MEGDR so it can decide when to use the hires data or not
4. Potentially add an offline command line argument so the simulation NEVER attempt to download.

I think this gives a good solution. What do you think?

**Comment by mokun** [comment]
> The LoRes images we can bundled as they are smaller if you want.

Yes agree. This is exactly what I would like to propose in this thread as suggested by the title. 

The point is that, in near future, we could use url to point to even more fine-grained res maps and height data set.

See this [commit](https://github.com/mars-sim/mars-sim/commit/b21dcf20b56448599dcc04f45cb25b08d40e9ad5) that I've just uploaded the low res (4 pixels per degree, 2 MB) height data img back to bundle with jar.

The problem is that the higher res height data (16 pixels per deg) is 32 MB and is currently not being bundled with jar. It needs to be downloaded from url. Without it, the simulation won't work.

**Comment by mokun** [comment]
> 1. Make MEGDR reader (or the calling class) lookup elevation directly form the image file. And cache recent values.

We are already reading directly from a double short int height array and not the img file. 

But here is the detail picture of what we've been doing in the past leading to the present. 

In the past before I joined in, we used to read height data from a jpg topo map. They had to cache the whole map image. 

1. A few years ago, we've switched to using MEGDR img file, instead of the old approach of reading a pixel value from a topo picture image.

2. The fact is that MEDGR img not a picture map but just a large data file containing a huge byte array of elevation data (from -6 km to 22 km) for the entire surface of Mars. They are arguably more precise in recording the elevation.

3. Recently, we've just eliminated the topo jpg map approach once and for all in favor of the MEGDR data approach. We've no longer performed the conversion of a color pixel to a height value, remember that you deleted those methods in `TerrainElevation` class ?

4. Currently, we only use MEGDR height data. However, I've started to doubt how well only using low res MEGDR height would serve us any good. 

5. Imagine that I'm at (0 N, 0 S) and say I'm at -3 km elevation. Since the low res MEGDR that bundles with jar has such a low res of 12 km per step of change of elevation. It's not until my rover travels to beyond 12 km that I would experience an elevation change. This is not good enough. 

6. But we could go back to reading compressed jpg topo map to get a value in between each adjacent pixel. So I could use a 5760x2880 topo map (can be compressed to only 10 MB) to supplement all those imprecise or `missing` elevation data that's being averaged away and not found in that low res MEGDR img file.

7. So my verdict that we by using both approach, we result in resolving elevation much better than using either one approach. 

Do you like we look into resurrecting the pixel value lookup for resolving elevation values by finding a good quality topo map ? 

We'll need to pick a map that has no label names on it.





But we never lookup the height dat

**Comment by mokun** [comment]
> Make MEGDR reader (or the calling class) lookup elevation directly form the image file. And cache recent values.

Yes we could look into how we do not have to store the entire map of elevation data into a big int array.

Currently, for fixed settlement, we no longer read from the array. We already cached the height value in each Settlement class.

But a Map lookup of elevation in `TerrainElevation` such one as below is very very huge and not as useful : 

```java
private Map<Coordinates, elevation> elevations = new HashMap<>();
```

Imagine a vehicle would travel 1000 km and at each frame, it would needs to compute its elevation value. The elevations Map would have hundreds of thousands of key-value pairs on it.

But say one day if we have a road network, it would make sense to have the values cache in some way when traveling on a designated route.

**Comment by mokun** [comment]
> 2. Make the downloading of files async with some callback notification.

Yes can we have that ?

Always, can we add a progress bar showing the percentage of downloading a large map or data file in the popup dialog box ? 

For slow internet connect, it would definitely help players understanding what's going on.

In future, the progress would be a good courtesy when we start making maps or height data that are couple hundreds MBs available for download.

**Comment by bevans2000** [comment]
> > The LoRes images we can bundled as they are smaller if you want.
> 
> Yes agree. This is exactly what I would like to propose in this thread as suggested by the title.
> 
> The point is that, in near future, we could use url to point to even more fine-grained res maps and height data set.
> 
> See this [commit](https://github.com/mars-sim/mars-sim/commit/b21dcf20b56448599dcc04f45cb25b08d40e9ad5) that I've just uploaded the low res (4 pixels per degree, 2 MB) height data img back to bundle with jar.
> 
> The problem is that the higher res height data (16 pixels per deg) is 32 MB and is currently not being bundled with jar. It needs to be downloaded from url. Without it, the simulation won't work.

Well it could work. We could run a flat Mars model and height for all coordinates would be zero if no data is available.
But the jar file is too big. It's not unreasonable to expect internet access. And once it has been download it will be reused

**Comment by bevans2000** [comment]
> > Make MEGDR reader (or the calling class) lookup elevation directly form the image file. And cache recent values.
> 
> Yes we could look into how we do not have to store the entire map of elevation data into a big int array.
> 
> Currently, for fixed settlement, we no longer read from the array. We already cached the height value in each Settlement class.
> 
> But a Map lookup of elevation in `TerrainElevation` such one as below is very very huge and not as useful :
> 
> ```java
> private Map<Coordinates, elevation> elevations = new HashMap<>();
> ```
> 
> Imagine a vehicle would travel 1000 km and at each frame, it would needs to compute its elevation value and store its value in this map. At the end, this elevations Map of one trip alone, it would have tens or hundreds of thousands of key-value pairs on it.
> 
> So reading from the int[][] array would be faster than going through the Coordinate class to find the phi and theta to compare if the coordinate is the same.
> 
> But say one day if we have a road network, it would make sense to have the values cache in some way when traveling on a designated route.

But if you consider the Vehicle trails it covers a very small percentage of the overall Mars surface so the double array I think will consume more data. If you want to use higher res terrain data then it can't be held in memory as it will be too big.

**Comment by bevans2000** [comment]
> > The LoRes images we can bundled as they are smaller if you want.
> 
> Yes agree. This is exactly what I would like to propose in this thread as suggested by the title.
> 
> The point is that, in near future, we could use url to point to even more fine-grained res maps and height data set.
> 
> See this [commit](https://github.com/mars-sim/mars-sim/commit/b21dcf20b56448599dcc04f45cb25b08d40e9ad5) that I've just uploaded the low res (4 pixels per degree, 2 MB) height data img back to bundle with jar.
> 
> The problem is that the higher res height data (16 pixels per deg) is 32 MB and is currently not being bundled with jar. It needs to be downloaded from url. Without it, the simulation won't work.

But in this commit we should have HiRes image of the default map layer bundled but I don't think it is.

**Comment by mokun** [comment]
> But in this commit we should have HiRes image of the default map layer bundled but I don't think it is.

But the hi res MEGDR img is 32 MB (compressible to 24 MB with zip and 18 MB with 7z). 

The thing is that 3.5.0 jar binary is only 37 MB. Now it would make a big jump in size.


So in case of low res, how about we supply it with a clean topo map image that we can extract elevation data ?

Say we can provide a 4096x2048 topo jpg map that's only 6.5 MB. 

This will provide close to 2.84x (1440 * 3 divided by 4096) better in order to supplement that low res 2 MB MEGDR img ?

**Comment by mokun** [comment]
> So in case of low res, how about we supply it with a clean topo map image that we can extract elevation data ?

> Say we can provide a 4096x2048 topo jpg map that's only 6.5 MB.

But having said that, we now have to go back to the same question of figuring out a sure way to convert a color pixel to a height value accurately.

**Comment by mokun** [comment]
@bevans2000 

I'm expanding the choice to hi, mid and low res for each map types. Committing soon.

**Comment by bevans2000** [comment]
So how would that work? Will it means more maps  bundled or are they downloadable?

**Comment by mokun** [comment]
What used to be high resolution is now mid res.

I've added the new hi res version of the Viking map and topo map to that `content` folder.

But there's a limit to the file size that I can uploaded there. 100 MB is the limit.

**Comment by bevans2000** [comment]
How big is the file if you zip it? I can add a unzip feature on the download easy. Having them zipped would actually make sense because we can also do it for the bundled maps.
Don't know why I didn't think about it originally

**Comment by mokun** [comment]
> Having them zipped would actually make sense because we can also do it for the bundled maps.

That's right ! Let's do it.

**Comment by bevans2000** [comment]
Ok, leave it with  me and I will have a go at weekend.

**Comment by mokun** [comment]
See this about uploading large files to GitHub :
https://docs.github.com/en/repositories/working-with-files/managing-large-files

**Comment by mokun** [comment]
### How big is the file if you zip it? 

I downloaded many of those GB image files from USGS website. 

For instances, the 3 GB topo file is at https://astrogeology.usgs.gov/search/map/Mars/GlobalSurveyor/MOLA/Mars_MGS_MOLA_ClrShade_merge_global_463m

However, in order to make it fit to our use, I have to downgrade its resolution in such a way that the size is reasonable enough for mars-sim.

So far, the largest file in the `content` folder is `Mars_Viking_MDIM21_ClrMosaic_1km.jpg`. It's 36 MB only.

I could create high resolution map files within 100 MB after being zipped up. 

This way, it can be easily be uploaded through the normal git commit process. 

If a file is beyond 100 MB, we'll need to use other tools. 

See also https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-large-files-on-github#about-size-limits-on-github.

**Comment by bevans2000** [comment]
Let's do the zipped support and see where it takes us

**Comment by bevans2000** [comment]
> Let's do the zipped support and see where it takes us

Had a look at this but the PNG & JPG file formats are already compressed so zipping them saves nothing.

**Comment by mokun** [comment]
> PNG & JPG file formats are already compressed so zipping them saves nothing.

Right. But compressing img files do reduce the size.

**Comment by bevans2000** [comment]
Yes about 20%. I've got the changed code now so I will give it a try.

**Comment by mokun** [comment]
Are you still working on this ?

I found out it's been loading the low res maps in all map types.

Not sure why yet. 

But if you're already debugging, I'll leave it to you.

**Comment by bevans2000** [comment]
Yes I should  plan to get back to this maybe tonight

**Comment by mokun** [comment]
The Java heap space problem is still there, when loading hi res map.

I use eclipse launcher to start with normal 1.5 GB memory and it is not enough.

So i need to change the Java switch to designate 2 GB to start JVM.

So when compiling jar, are you comfortable with designating 2GB to start ? I haven't done that in case of starting in jar.

What's the solution ?

**Comment by bevans2000** [comment]
Solution is to reduce the memory consumption for images. I'm confident I can reduce it for MEGDR, just need to find time to do it.

**Comment by mokun** [comment]
Currently we sonr use much memory doe elevation yet but could be better if we start using the quad image map for elevation.

On the other hand, the large double int array generated from fhe hi res map is the real culprit in this case.

It just dawn on me that we can simply create a smaller size array with just half the map size, because we can only project half of the map onto 2D map box and see at most half the globe only.

So there is no need to generate a big double array encompassing the whole globe.

**Comment by bevans2000** [comment]
OK, I'll look at the MEGDR and elevation array then and leave the image maps to you

**Comment by bevans2000** [comment]
So I have done a rework of the MEGDR logic. Now the image file is  accessed via Memory mapping logic which means it is requires no Java heap to hold the data. The data values are read on demand directly form the file. There is a caching logic layer in place to hold 1000 values for upto 5 minutes.
The old array approach needs 77Mb of memory but the memory mapped approached needs 2Mb. 
There is also a second approach that uses the Random file access to get the bytes directly out of a file.
I have add the selection of the reader to be defined in the mapdata.properties file.
I just need to create a PR to get it in the master branch

**Comment by mokun** [comment]
> The old array approach needs 77Mb of memory but the memory mapped approached needs 2Mb.

I see this as a really good start. Next step is to enable online download of higher resolution elevation data set. Since it won't need to create an 2D int array (is that true ?), this would save much more heap memory.

**Comment by bevans2000** [comment]
> > The old array approach needs 77Mb of memory but the memory mapped approached needs 2Mb.
> 
> I see this as a really good start. Next step is to enable online download of higher resolution elevation data set. Since it won't need to create an 2D int array (is that true ?), this would save much more heap memory.

Yes there is no need for the 2D array now so a larger resolution should be possible.

**Comment by mokun** [comment]
Ready to ommit or pull ?

**Comment by bevans2000** [comment]
Iwill create a pull later today after work

**Comment by mokun** [comment]
Awesome change !

So now, if I'm to change it to downloading higher res elevation img file, we'll need to convert the downloading process to run on a thread. Shall we ? 

If not, it would look as if the whole simulation is halted and there's no guarantee if the download is successful. 

If successful, it will need to tag the new img file as the one it will use.

If failed or time out, how do we handle it the exception ?

**Comment by bevans2000** [comment]
Yes this change should help with the resolution.
Remember it will only download the image once per installation. Using asynchronous loading of remote content is not that easy because the calling class will be using a blocking approach, so if the image is not available it cannot do anything.

**Comment by mokun** [comment]
I was reading at some of the async lib out there such as [1](https://ribasco.github.io/async-gamequery-lib/). 

> Using asynchronous loading of remote content is not that easy

So what is the hardest part ?

I suppose we can let player know that the newly downloaded elevation data set or surface map won't be available until saving, exiting, and reloading mars-sim

**Comment by bevans2000** [comment]
The terrain is downloaded as part of the simulation build so the UI hasn't even started nor is there the chance to ask the user.

**Comment by bevans2000** [comment]
We could get the Filelocator class to return Future<File> instead of a File but it wouldn't help with the elevation because it is needed to create a Settlement's terrain profile at simulation building. We could change that to be lazy loading I guess which could work.

**Comment by mokun** [comment]
> We could change that to be lazy loading I guess which could work.

That will be awesome.

**Comment by mokun** [comment]
A question, in terms of caching the elevation values.

Say a vehicle travels from (0.0001 N, 0.0001 S)  to (0.0002 N, 0.0002 S).

Since the lat and lon values are currently up to 4 decimal places, it will be two distinct coordinates.

However, the elevation values on both of them would be the same.

That's because the current resolution of the elevation data is only 16 pixels per degree, or   0.0625 deg per pixel, which is translated to 3.705 km per pixel at the equator.

The difference between 0.001 and 0.002 deg in latitude makes no difference and will load up the same elevation value.

How would you make the caching more efficient in light of this ?

```
//	LEVEL 1 :
//  File name: megt90n000eb.img
//	Pixels: 5760 x 2880
//	Resolution: 16 pixels per degree (or 0.0625by 0.0625 degrees)
//	Scale: 3.705 km per pixel (=1/16)	
//  File size: 32.4 MB
```

**Comment by mokun** [comment]
btw, I implement the use of xz compression/decompression here because it save a lot of space. e.g. original : 32.4 MB. Compress to 17.4 MB (in xz) versus 23.3 MB (in zip).

**Comment by bevans2000** [comment]
> How would you make the caching more efficient in light of this ?
> 
The caching g handles this because it catches at the file index level and not the coordinates. So in your example both terrain requests would get the same value out of the cache because they resolve to the same index. Hence if the file resolution changes, the caching still works efficiently.

**Comment by bevans2000** [comment]
> btw, I implement the use of xz compression/decompression here because it save a lot of space. e.g. original : 32.4 MB. Compress to 17.4 MB (in xz) versus 23.3 MB (in zip).

Great answer using XZ compression. We can use that for other files if needed as the JPEG file do not compress at the moment using ZIP. If we correct FileLocator to support XZ files generically we could use it for the JPEG files as well.

**Comment by mokun** [comment]
> FileLocator to support XZ files generically we could use it for the JPEG files as well.

But xz wont work well for jpg either. 

I could try to see if xz works on some original maps that are in png or tiff.

**Comment by bevans2000** [comment]
I want to clean FileLocator as it's got a bit messy

**Comment by mokun** [comment]
> I want to clean FileLocator as it's got a bit messy

Sure. Please do go ahead relocating whatever chucks of codes that don't belong.

**Comment by mokun** [comment]
> Make the downloading of files async with some callback notification.

Now that downloading surface map are async, can we make downloading elevation data async also ?

**Comment by mokun** [comment]
Async download is almost a prerequisite for the tile loading of elevation data.

**Comment by bevans2000** [comment]
Yes in theory but there is an important difference between MapData and MEDGRreader. For terrain elevation it is a blocking method with no change to change it to async. So it will have to start with the local lores file and do a background download of hi res, then flip over once locally available. 
The problem might be memory consumption having the 2 terrain data in memory for the transition period.
It can be done but tiling this would be better. It could only keep the required data in memory.
Now considering tiling MapData there is a lot of the same logic so we should be able to combine it to a common class.
